# Data Cleaning

In [1]:
import pandas as pd
import numpy as np


In [2]:
smartphones=pd.read_csv('/content/drive/MyDrive/dataset/smartphones - smartphones.csv')

In [3]:
smartphones['card'].value_counts()

,count
card,
"Memory Card Supported, upto 1 TB",171
Memory Card Not Supported,112
Android v12,107
"Memory Card Supported, upto 512 GB",105
"Memory Card (Hybrid), upto 1 TB",91
...,...
"Memory Card Supported, upto 48 GB",1
"Memory Card (Hybrid), upto 2 TB",1
HarmonyOS,1


# Data Accessing
# **Quality Issues**

1. **model** - some brands are written diiferently like OPPO(upper and lower) in model column `consistency`
2. **price** - has unneccesary '₹' `validity`
3. **price** - has ',' between numbers `validity`
4. **price** - phone Namotel has a price of 99 `accuracy`
5. **ratings** - missing values `completeness` also rating in whole number i need float like 8.9
6. **processor** - has some incorrect values for some samsung phones(row # -642,647,649,659,667,701,750,759,819,859,883,884,919,927,929,932,1002) `validity`
7. There is ipod on row 756 `validity`
8. **memory** - incorrect values in rows (441,485,534,553,584,610,613,642,647,649,659,667,701,750,759,819,859,884,919,927,929,932,990,1002) `validity`
14 row= no gb there is battery given
9. **battery** - incorrect values in rows(113,151,309,365,378,441,450,553,584,610,613,630,642,647,649,659,667,701,750,756,759,764,819,855,859,884,915,916,927,929,932,990,1002) `validity`
10. **display** - sometimes frequency is not available `completeness`
11. **display** - incorrect values in rows(378,441,450,553,584,610,613,630,642,647,649,659,667,701,750,759,764,819,859,884,915,916,927,929,932,990,1002) `validity`
12. certain phones are foldable and the info is scattered `validity`
13. **camera** - words like Dual, Triple and Quad are used to represent number of cameras and front and rear cameras are separated by '&'
14. **camera** - problem with rows (100,113,151,157,161,238,273,308,309,323,324,365,367,378,394,441,450,484,506,534,553,571,572,575,584,610,613,615,630,642,647,649,659,667,684,687,705,711,723,728,750,756,759,764,792,819,846,854,855,858,883,884,896,915,916,927,929,932,945,956,990,995,1002,1016
) `validity`
15. **card** - sometimes contains info about os and camera `validity`
16. **os** - sometimes contains info about bluetooth and fm radio `validity`
17. **os** - issue with rows (324,378) `validity`
18. **os** - sometimes contains os version name like lollipop `consistency`
19. missing values in camera, card and os `completeness`
 missing values in camera card and os
20. datatype  of price and rating is incorrect `validity`
data type of price and rating incorrect



### Tidiness Issues

1. **sim** - can be split into 3 cols has_5g, has_NFC, has_IR_Blaster
2. **ram** - can be split into 2 cols RAM and ROM
3. **processor** - can be split into processor name, cores and cpu speed.
4. **battery** - can be split into battery capacity, fast_charging_available
5. **display** - can be split into size, resolution_width, resolution_height and frequency
6. **camera** - can be split into front and rear camera
7. **card** - can be split into supported, extended_upto

In [4]:
smartphones.head(2)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12


In [5]:
smartphones.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   model      1020 non-null   object 
 1   price      1020 non-null   object 
 2   rating     879 non-null    float64
 3   sim        1020 non-null   object 
 4   processor  1020 non-null   object 
 5   ram        1020 non-null   object 
 6   battery    1020 non-null   object 
 7   display    1020 non-null   object 
 8   camera     1019 non-null   object 
 9   card       1013 non-null   object 
 10  os         1003 non-null   object 
dtypes: float64(1), object(10)
memory usage: 87.8+ KB


In [6]:
smartphones.duplicated().sum()

np.int64(0)

In [7]:
smartphones.isnull().sum()

,0
model,0
price,0
rating,141
sim,0
processor,0
ram,0
battery,0
display,0
camera,1
card,7


# cleaning
quality issue
completeness,inconsistencies, accuracy,validity

In [8]:
smartphones.head()

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13


# validity issue
1. price - has unneccesary '₹' validity
2. price - has ',' between numbers validity

In [9]:
smartphones_copy=smartphones.copy()

In [10]:
smartphones_copy.head()

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13


In [11]:
smartphones_copy['price']=smartphones_copy['price'].str.replace('₹','').str.replace(',','').astype('int')

#1. index in excel starting from 2 but in data here index starting from 0



In [12]:
smartphones_copy=smartphones_copy.reset_index()

In [13]:
smartphones_copy['index']=smartphones_copy['index']+2

In [14]:
smartphones_copy

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,2,OnePlus 11 5G,54999,89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,3,OnePlus Nord CE 2 Lite 5G,19989,81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,4,Samsung Galaxy A14 5G,16499,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,5,Motorola Moto G62 5G,14999,81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,6,Realme 10 Pro Plus,24999,82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13
...,...,...,...,...,...,...,...,...,...,...,...,...
1015,1017,Motorola Moto Edge S30 Pro,34990,83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 68.2W Fast Charging,"6.67 inches, 1080 x 2460 px, 120 Hz Display wi...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
1016,1018,Honor X8 5G,14990,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 480+, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 22.5W Fast Charging,"6.5 inches, 720 x 1600 px Display with Water D...",48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...,"Memory Card Supported, upto 1 TB",Android v11
1017,1019,POCO X4 GT 5G (8GB RAM + 256GB),28990,85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Dimensity 8100, Octa Core, 2.85 GHz Processor","8 GB RAM, 256 GB inbuilt",5080 mAh Battery with 67W Fast Charging,"6.6 inches, 1080 x 2460 px, 144 Hz Display wit...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Memory Card Not Supported,Android v12
1018,1020,Motorola Moto G91 5G,19990,80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.8 inches, 1080 x 2400 px Display with Punch ...",108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...,"Memory Card Supported, upto 1 TB",Android v12


# problametic rows

In [15]:
processor_rows=set((642,647,649,659,667,701,750,759,819,859,883,884,919,927,929,932,1002))
ram_rows=set((441,485,534,553,584,610,613,642,647,649,659,667,701,750,759,819,859,884,919,927,929,932,990,1002))
battery_rows=set((113,151,309,365,378,441,450,553,584,610,613,630,642,647,649,659,667,701,750,756,759,764,819,855,859,884,915,916,927,929,932,990,1002))
display_rows=set((378,441,450,553,584,610,613,630,642,647,649,659,667,701,750,759,764,819,859,884,915,916,927,929,932,990,1002))
camera_rows=set((100,113,151,157,161,238,273,308,309,323,324,365,367,378,394,441,450,484,506,534,553,571,572,575,584,610,613,615,630,642,647,649,659,667,684,687,705,711,723,728,750,756,759,764,792,819,846,854,855,858,883,884,896,915,916,927,929,932,945,956,990,995,1002,1016 ))

In [16]:
smartphones_copy[smartphones_copy['index'].isin(processor_rows|ram_rows|battery_rows|display_rows|camera_rows)]# atleast one  coloumn problem

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
98,100,Vivo X Fold 5G,106990,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","12 GB RAM, 256 GB inbuilt",4600 mAh Battery with 66W Fast Charging,"8.03 inches, 1916 x 2160 px, 120 Hz Display",Foldable Display,50 MP Quad Rear & 16 MP Front Camera,Android v12
111,113,Apple iPhone 12,51999,74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
149,151,Apple iPhone 12 Mini,40999,74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
155,157,Nokia 2780 Flip,4990,NaN,"Dual Sim, 3G, 4G, Wi-Fi","Snapdragon QM215, Quad Core, 1.3 GHz Processor","4 GB RAM, 512 MB inbuilt",1450 mAh Battery,"2.7 inches, 240 x 320 px Display",Dual Display,5 MP Rear Camera,"Memory Card Supported, upto 32 GB"
159,161,Oppo Find N2 5G,94990,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",4520 mAh Battery with 67W Fast Charging,"7.1 inches, 1792 x 1920 px, 120 Hz Display wit...","Foldable Display, Dual Display",50 MP + 48 MP + 32 MP Triple Rear & 32 MP + 32...,Memory Card Not Supported
...,...,...,...,...,...,...,...,...,...,...,...,...
954,956,Vivo X Fold 5G (12GB RAM + 512GB),118990,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","12 GB RAM, 512 GB inbuilt",4600 mAh Battery with 66W Fast Charging,"8.03 inches, 1916 x 2160 px, 120 Hz Display",Foldable Display,50 MP Quad Rear & 16 MP Front Camera,Android v12
988,990,Nokia 5310 Dual Sim,3399,NaN,Dual Sim,"8 MB RAM, 16 MB inbuilt",1200 mAh Battery,"2.4 inches, 240 x 320 px Display",0.3 MP Rear Camera,"Memory Card Supported, upto 32 GB",Bluetooth,Browser
993,995,Huawei Mate X,169000,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Kirin 990, Octa Core, 2.86 GHz Processor","8 GB RAM, 512 GB inbuilt",4500 mAh Battery with 55W Fast Charging,"8 inches, 2200 x 2480 px Display",Foldable Display,48 MP Quad Rear Camera,"Memory Card (Hybrid), upto 256 GB"
1000,1002,XTouch F40 Flip,1999,NaN,Dual Sim,No 3G,No Wifi,"32 MB RAM, 32 MB inbuilt",800 mAh Battery,"1.77 inches, 240 x 320 px Display",Dual Display,1.3 MP Rear Camera


In [17]:
smartphones_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   index      1020 non-null   int64  
 1   model      1020 non-null   object 
 2   price      1020 non-null   int64  
 3   rating     879 non-null    float64
 4   sim        1020 non-null   object 
 5   processor  1020 non-null   object 
 6   ram        1020 non-null   object 
 7   battery    1020 non-null   object 
 8   display    1020 non-null   object 
 9   camera     1019 non-null   object 
 10  card       1013 non-null   object 
 11  os         1003 non-null   object 
dtypes: float64(1), int64(2), object(9)
memory usage: 95.8+ KB


In [18]:
smartphones_copy[smartphones_copy['index'].isin(processor_rows&ram_rows&battery_rows&display_rows&camera_rows)]#every rows coloumn  has problem #feature phone

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
640,642,Nokia 105 Plus,1299,NaN,Dual Sim,"4 MB RAM, 4 MB inbuilt",800 mAh Battery,"1.77 inches, 128 x 160 px Display",No Rear Camera,"Memory Card Supported, upto 32 GB",Bluetooth,NaN
645,647,Nokia 2760 Flip,5490,NaN,"Dual Sim, 3G, 4G, Wi-Fi",1450 mAh Battery,"3.6 inches, 240 x 320 px Display",5 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 32 GB",Kaios v3.0,Bluetooth,NaN
647,649,Motorola Moto A10,1339,NaN,Dual Sim,"4 MB RAM, 4 MB inbuilt",1750 mAh Battery,"1.8 inches, 160 x 128 px Display",No Rear Camera,"Memory Card Supported, upto 32 GB",NaN,NaN
657,659,Zanco Tiny T1,2799,NaN,Single Sim,"32 MB RAM, 32 MB inbuilt",200 mAh Battery,"0.49 inches, 64 x 32 px Display",No Rear Camera,No FM Radio,Bluetooth,NaN
665,667,itel it2163S,958,NaN,Dual Sim,"4 MB RAM, 4 MB inbuilt",1200 mAh Battery,"1.8 inches, 160 x 128 px Display",No Rear Camera,"Memory Card Supported, upto 32 GB",Bluetooth,NaN
748,750,Nokia 400 4G,3290,NaN,"Dual Sim, 4G, VoLTE, Wi-Fi",2000 mAh Battery,"2.4 inches, 240 x 320 px Display",0.3 MP Rear & 0.3 MP Front Camera,"Memory Card Supported, upto 64 GB",Bluetooth,Browser,NaN
757,759,Karbonn KU3i,995,NaN,Dual Sim,"52 MB RAM, 32 MB inbuilt",1000 mAh Battery,"1.8 inches, 128 x 160 px Display",No Rear Camera,"Memory Card Supported, upto 16 GB",Bluetooth,NaN
817,819,itel Magic X,2239,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi",No 3G,T117,"48 MB RAM, 128 MB inbuilt",1200 mAh Battery,"2.4 inches, 240 x 320 px Display",1.3 MP Rear Camera,"Memory Card Supported, upto 64 GB"
882,884,Nokia 5710 XpressAudio,4799,NaN,"Dual Sim, 3G, 4G",No Wifi,Unisoc T107,"48 MB RAM, 128 MB inbuilt",1450 mAh Battery,"2.4 inches, 240 x 320 px Display",0.3 MP Rear Camera,"Memory Card Supported, upto 32 GB"
925,927,Nokia 3310 4G,3999,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","256 MB RAM, 512 MB inbuilt",1200 mAh Battery,"2.4 inches, 240 x 320 px Display",2 MP Rear Camera,"Memory Card Supported, upto 32 GB",Bluetooth,Browser


In [19]:
smartphones_copy[smartphones_copy['index'].isin(processor_rows&ram_rows&battery_rows&display_rows&camera_rows)]['price']

,price
640,1299
645,5490
647,1339
657,2799
665,958
748,3290
757,995
817,2239
882,4799
925,3999


In [20]:
smartphones_copy=smartphones_copy[smartphones_copy['price']>3400]#dropping out because less than 3000 are not smart ones
smartphones_copy

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,2,OnePlus 11 5G,54999,89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,3,OnePlus Nord CE 2 Lite 5G,19989,81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,4,Samsung Galaxy A14 5G,16499,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,5,Motorola Moto G62 5G,14999,81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,6,Realme 10 Pro Plus,24999,82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13
...,...,...,...,...,...,...,...,...,...,...,...,...
1015,1017,Motorola Moto Edge S30 Pro,34990,83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 68.2W Fast Charging,"6.67 inches, 1080 x 2460 px, 120 Hz Display wi...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
1016,1018,Honor X8 5G,14990,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 480+, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 22.5W Fast Charging,"6.5 inches, 720 x 1600 px Display with Water D...",48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...,"Memory Card Supported, upto 1 TB",Android v11
1017,1019,POCO X4 GT 5G (8GB RAM + 256GB),28990,85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Dimensity 8100, Octa Core, 2.85 GHz Processor","8 GB RAM, 256 GB inbuilt",5080 mAh Battery with 67W Fast Charging,"6.6 inches, 1080 x 2460 px, 144 Hz Display wit...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Memory Card Not Supported,Android v12
1018,1020,Motorola Moto G91 5G,19990,80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.8 inches, 1080 x 2400 px Display with Punch ...",108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...,"Memory Card Supported, upto 1 TB",Android v12


In [21]:
smartphones_copy[smartphones_copy['index'].isin(processor_rows|ram_rows|battery_rows|display_rows|camera_rows)]

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
98,100,Vivo X Fold 5G,106990,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","12 GB RAM, 256 GB inbuilt",4600 mAh Battery with 66W Fast Charging,"8.03 inches, 1916 x 2160 px, 120 Hz Display",Foldable Display,50 MP Quad Rear & 16 MP Front Camera,Android v12
111,113,Apple iPhone 12,51999,74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
149,151,Apple iPhone 12 Mini,40999,74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
155,157,Nokia 2780 Flip,4990,NaN,"Dual Sim, 3G, 4G, Wi-Fi","Snapdragon QM215, Quad Core, 1.3 GHz Processor","4 GB RAM, 512 MB inbuilt",1450 mAh Battery,"2.7 inches, 240 x 320 px Display",Dual Display,5 MP Rear Camera,"Memory Card Supported, upto 32 GB"
159,161,Oppo Find N2 5G,94990,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",4520 mAh Battery with 67W Fast Charging,"7.1 inches, 1792 x 1920 px, 120 Hz Display wit...","Foldable Display, Dual Display",50 MP + 48 MP + 32 MP Triple Rear & 32 MP + 32...,Memory Card Not Supported
236,238,Xiaomi Mix Fold 2 5G,106990,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Snapdragon 8+ Gen1 , Octa Core, 3.2 GHz Proce...","12 GB RAM, 256 GB inbuilt",4500 mAh Battery with 67W Fast Charging,"8.02 inches, 1914 x 2160 px, 120 Hz Display wi...","Foldable Display, Dual Display",50 MP + 13 MP + 8 MP Triple Rear & 20 MP Front...,Android v12
271,273,Nokia 2720 V Flip,6199,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 205 , Dual Core, 1.1 GHz Processor","512 MB RAM, 4 GB inbuilt",1500 mAh Battery,"2.8 inches, 240 x 320 px Display",Dual Display,2 MP Rear Camera,Memory Card Supported
306,308,Samsung Galaxy Z Flip 3,69999,84.0,"Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 888, Octa Core, 2.84 GHz Processor","8 GB RAM, 128 GB inbuilt",3300 mAh Battery with 15W Fast Charging,"6.7 inches, 1080 x 2640 px, 120 Hz Display wit...","Foldable Display, Dual Display",12 MP + 12 MP Dual Rear & 10 MP Front Camera,Memory Card Not Supported
307,309,Apple iPhone 12 (128GB),55999,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
321,323,Samsung Galaxy Z Fold 4,154998,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",4400 mAh Battery with 25W Fast Charging,"7.6 inches, 1812 x 2176 px, 120 Hz Display wit...","Foldable Display, Dual Display",50 MP + 12 MP + 10 MP Triple Rear & 10 MP + 4 ...,Android v12


In [22]:
smartphones_copy[smartphones_copy['index'].isin(processor_rows&ram_rows&battery_rows&display_rows&camera_rows)]

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
645,647,Nokia 2760 Flip,5490,NaN,"Dual Sim, 3G, 4G, Wi-Fi",1450 mAh Battery,"3.6 inches, 240 x 320 px Display",5 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 32 GB",Kaios v3.0,Bluetooth,NaN
882,884,Nokia 5710 XpressAudio,4799,NaN,"Dual Sim, 3G, 4G",No Wifi,Unisoc T107,"48 MB RAM, 128 MB inbuilt",1450 mAh Battery,"2.4 inches, 240 x 320 px Display",0.3 MP Rear Camera,"Memory Card Supported, upto 32 GB"
925,927,Nokia 3310 4G,3999,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","256 MB RAM, 512 MB inbuilt",1200 mAh Battery,"2.4 inches, 240 x 320 px Display",2 MP Rear Camera,"Memory Card Supported, upto 32 GB",Bluetooth,Browser


#processor - has some incorrect values for some samsung phones(row # -642,647,649,659,667,701,750,759,819,859,883,884,919,927,929,932,1002) validity
There is ipod on row 756 validity

In [23]:
smartphones_copy[smartphones_copy['index'].isin(processor_rows)]#qll are feature

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
645,647,Nokia 2760 Flip,5490,NaN,"Dual Sim, 3G, 4G, Wi-Fi",1450 mAh Battery,"3.6 inches, 240 x 320 px Display",5 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 32 GB",Kaios v3.0,Bluetooth,NaN
857,859,LG Folder 2,11999,NaN,"Single Sim, 3G, 4G, Wi-Fi","1 GB RAM, 8 GB inbuilt",1470 mAh Battery,"2.8 inches, 240 x 320 px Display",2 MP Rear Camera,Memory Card Supported,Bluetooth,NaN
882,884,Nokia 5710 XpressAudio,4799,NaN,"Dual Sim, 3G, 4G",No Wifi,Unisoc T107,"48 MB RAM, 128 MB inbuilt",1450 mAh Battery,"2.4 inches, 240 x 320 px Display",0.3 MP Rear Camera,"Memory Card Supported, upto 32 GB"
925,927,Nokia 3310 4G,3999,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","256 MB RAM, 512 MB inbuilt",1200 mAh Battery,"2.4 inches, 240 x 320 px Display",2 MP Rear Camera,"Memory Card Supported, upto 32 GB",Bluetooth,Browser


In [24]:
len(processor_rows)

17

In [25]:
smartphones_copy.drop([645,857,882,925],inplace=True)

<ipython-input-25-fcac0713b134>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  smartphones_copy.drop([645,857,882,925],inplace=True)


# ram memory
 - incorrect values in rows (441,485,534,553,584,610,613,642,647,649,659,667,701,750,759,819,859,884,919,927,929,932,990,1002) validity 14 row= no gb there is battery given

In [26]:
smartphones_copy[smartphones_copy['index'].isin(ram_rows)]

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
439,441,Apple iPhone SE 3 2022,43900,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor",64 GB inbuilt,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v15,No FM Radio
483,485,Huawei Mate 50 RS Porsche Design,239999,81.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC, IR Blaster","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor",512 GB inbuilt,4700 mAh Battery with 66W Fast Charging,"6.74 inches, 1212 x 2616 px, 120 Hz Display",50 MP + 48 MP + 13 MP Triple Rear & 13 MP Fron...,"Memory Card (Hybrid), upto 256 GB",Hongmeng OS v3.0
582,584,Nokia 8210 4G,3749,NaN,"Dual Sim, 3G, 4G",No Wifi,Unisoc T107,"48 MB RAM, 128 MB inbuilt",1450 mAh Battery,"2.8 inches, 240 x 320 px Display",0.3 MP Rear Camera,"Memory Card Supported, upto 32 GB"


In [27]:
smartphones_copy.drop(582,inplace=True)

<ipython-input-27-e87455e7c1d5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  smartphones_copy.drop(582,inplace=True)


#battery -
incorrect values in rows(113,151,309,365,378,441,450,553,584,610,613,630,642,647,649,659,667,701,750,756,759,764,819,855,859,884,915,916,927,929,932,990,1002) validity

In [28]:
smartphones_copy.drop([376,754],inplace=True)

<ipython-input-28-9a52a5d9fb46>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  smartphones_copy.drop([376,754],inplace=True)


In [29]:
temp_df=smartphones_copy[smartphones_copy['index'].isin(battery_rows)]
temp_df

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
111,113,Apple iPhone 12,51999,74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
149,151,Apple iPhone 12 Mini,40999,74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
307,309,Apple iPhone 12 (128GB),55999,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
363,365,Apple iPhone 12 Mini (128GB),45999,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
439,441,Apple iPhone SE 3 2022,43900,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor",64 GB inbuilt,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v15,No FM Radio
448,450,Apple iPhone 15 Pro,130990,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC",Bionic A16,"8 GB RAM, 128 GB inbuilt","6.06 inches, 1170 x 2532 px, 120 Hz Display wi...",50 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15,No FM Radio
628,630,Apple iPhone 12 Pro (512GB),139900,80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","6 GB RAM, 512 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v14.0,No FM Radio
762,764,Apple iPhone SE 4,49990,60.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor",64 GB inbuilt,"6.1 inches, 750 x 1580 px Display",12 MP Rear & 10.8 MP Front Camera,Memory Card Not Supported,iOS v16,No FM Radio
853,855,Apple iPhone 12 Pro (256GB),119900,80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","6 GB RAM, 256 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v14.0,No FM Radio
913,915,Apple iPhone 12 Mini (256GB),55999,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 256 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio


In [30]:
x=temp_df.iloc[:,7:].shift(1,axis=1).values
x

array([[None,
        '6.1 inches, 1170\u2009x\u20092532\u2009px Display with Large Notch',
        '12\u2009MP + 12\u2009MP Dual Rear & 12\u2009MP Front Camera',
        'Memory Card Not Supported', 'iOS v14'],
       [None, '5.4 inches, 1080\u2009x\u20092340\u2009px Display',
        '12\u2009MP + 12\u2009MP Dual Rear & 12\u2009MP Front Camera',
        'Memory Card Not Supported', 'iOS v14'],
       [None,
        '6.1 inches, 1170\u2009x\u20092532\u2009px Display with Large Notch',
        '12\u2009MP + 12\u2009MP Dual Rear & 12\u2009MP Front Camera',
        'Memory Card Not Supported', 'iOS v14'],
       [None, '5.4 inches, 1080\u2009x\u20092340\u2009px Display',
        '12\u2009MP + 12\u2009MP Dual Rear & 12\u2009MP Front Camera',
        'Memory Card Not Supported', 'iOS v14'],
       [None, '4.7 inches, 750\u2009x\u20091334\u2009px Display',
        '12\u2009MP Rear & 7\u2009MP Front Camera',
        'Memory Card Not Supported', 'iOS v15'],
       [None,
        '6.06 inches,

In [31]:
x_index=temp_df.index

In [32]:
smartphones_copy.loc[x_index,temp_df.columns[7:]]=x

# display -
incorrect values in rows(378,441,450,553,584,610,613,630,642,647,649,659,667,701,750,759,764,819,859,884,915,916,927,929,932,990,1002) validity

In [33]:
smartphones_copy[smartphones_copy['index'].isin(display_rows)]

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
439,441,Apple iPhone SE 3 2022,43900,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor",64 GB inbuilt,None,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v15
448,450,Apple iPhone 15 Pro,130990,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC",Bionic A16,"8 GB RAM, 128 GB inbuilt",None,"6.06 inches, 1170 x 2532 px, 120 Hz Display wi...",50 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15
628,630,Apple iPhone 12 Pro (512GB),139900,80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","6 GB RAM, 512 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v14.0
762,764,Apple iPhone SE 4,49990,60.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor",64 GB inbuilt,None,"6.1 inches, 750 x 1580 px Display",12 MP Rear & 10.8 MP Front Camera,Memory Card Not Supported,iOS v16
913,915,Apple iPhone 12 Mini (256GB),55999,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 256 GB inbuilt",None,"5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
914,916,Apple iPhone 12 (256GB),67999,76.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 256 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14


# camera -
 problem with rows (100,113,151,157,161,238,273,308,309,323,324,365,367,378,394,441,450,484,506,534,553,571,572,575,584,610,613,615,630,642,647,649,659,667,684,687,705,711,723,728,750,756,759,764,792,819,846,854,855,858,883,884,896,915,916,927,929,932,945,956,990,995,1002,1016 )

In [34]:
len(camera_rows)

64

155,

In [35]:
smartphones_copy[smartphones_copy['index'].isin(camera_rows)]

,index,model,price,rating,sim,processor,ram,battery,display,camera,card,os
98,100,Vivo X Fold 5G,106990,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","12 GB RAM, 256 GB inbuilt",4600 mAh Battery with 66W Fast Charging,"8.03 inches, 1916 x 2160 px, 120 Hz Display",Foldable Display,50 MP Quad Rear & 16 MP Front Camera,Android v12
111,113,Apple iPhone 12,51999,74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
149,151,Apple iPhone 12 Mini,40999,74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt",None,"5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
155,157,Nokia 2780 Flip,4990,NaN,"Dual Sim, 3G, 4G, Wi-Fi","Snapdragon QM215, Quad Core, 1.3 GHz Processor","4 GB RAM, 512 MB inbuilt",1450 mAh Battery,"2.7 inches, 240 x 320 px Display",Dual Display,5 MP Rear Camera,"Memory Card Supported, upto 32 GB"
159,161,Oppo Find N2 5G,94990,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",4520 mAh Battery with 67W Fast Charging,"7.1 inches, 1792 x 1920 px, 120 Hz Display wit...","Foldable Display, Dual Display",50 MP + 48 MP + 32 MP Triple Rear & 32 MP + 32...,Memory Card Not Supported
236,238,Xiaomi Mix Fold 2 5G,106990,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Snapdragon 8+ Gen1 , Octa Core, 3.2 GHz Proce...","12 GB RAM, 256 GB inbuilt",4500 mAh Battery with 67W Fast Charging,"8.02 inches, 1914 x 2160 px, 120 Hz Display wi...","Foldable Display, Dual Display",50 MP + 13 MP + 8 MP Triple Rear & 20 MP Front...,Android v12
271,273,Nokia 2720 V Flip,6199,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 205 , Dual Core, 1.1 GHz Processor","512 MB RAM, 4 GB inbuilt",1500 mAh Battery,"2.8 inches, 240 x 320 px Display",Dual Display,2 MP Rear Camera,Memory Card Supported
306,308,Samsung Galaxy Z Flip 3,69999,84.0,"Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 888, Octa Core, 2.84 GHz Processor","8 GB RAM, 128 GB inbuilt",3300 mAh Battery with 15W Fast Charging,"6.7 inches, 1080 x 2640 px, 120 Hz Display wit...","Foldable Display, Dual Display",12 MP + 12 MP Dual Rear & 10 MP Front Camera,Memory Card Not Supported
307,309,Apple iPhone 12 (128GB),55999,75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
321,323,Samsung Galaxy Z Fold 4,154998,NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",4400 mAh Battery with 25W Fast Charging,"7.6 inches, 1812 x 2176 px, 120 Hz Display wit...","Foldable Display, Dual Display",50 MP + 12 MP + 10 MP Triple Rear & 10 MP + 4 ...,Android v12


In [36]:
smartphones_copy.drop([155,271],inplace=True)

<ipython-input-36-865e26ef3e42>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  smartphones_copy.drop([155,271],inplace=True)


In [ ]:
temp_df=smartphones_copy[smartphones_copy['index'].isin(camera_rows)]


In [ ]:
temp_df

In [ ]:
temp_df=temp_df[~temp_df['camera'].str.contains('MP')]#problematic

In [ ]:
smartphones_copy.loc[temp_df.index,'camera']=temp_df['card'].values

In [ ]:
temp_df

In [ ]:
smartphones_copy[smartphones_copy['index'].isin(camera_rows)]

# card
card - sometimes contains info about os and camera validity

In [ ]:
smartphones_copy

In [ ]:
temp_df=smartphones_copy[smartphones_copy['card'].str.contains('OS',na=False)]

In [ ]:
x=temp_df.iloc[:,10:].shift(1,axis=1).values

In [ ]:
x_index=temp_df.index

In [ ]:
smartphones_copy.loc[x_index,temp_df.columns[10:]]=x

In [ ]:
temp_df=smartphones_copy[smartphones_copy['card'].str.contains('Android',na=False)]


In [ ]:
x=temp_df.iloc[:,10:].shift(1,axis=1).values

In [ ]:
x

In [ ]:
x_index=temp_df.index

In [ ]:
smartphones_copy.loc[x_index,temp_df.columns[10:]]=x

In [ ]:
temp_df.index

In [ ]:
smartphones_copy.loc[temp_df.index,'card']=temp_df['card']

In [ ]:
temp_df=smartphones_copy[smartphones_copy['card'].str.contains('MP',na=False)]

In [ ]:
smartphones_copy.loc[temp_df.index,'card']='Memory Card Supported'

In [ ]:
smartphones_copy['card'].value_counts()

In [ ]:
temp_df=smartphones_copy[~smartphones_copy['card'].str.contains('Memory Card',na=False)]

In [ ]:
temp_df.index

In [ ]:
temp_df=smartphones_copy[smartphones_copy['card']=='Memory Card not Supported']

In [ ]:
smartphones_copy.loc[temp_df.index,'card']='Memory Card Not Supported'

# os

In [ ]:
temp_df=smartphones_copy[smartphones_copy['os'].str.contains('Memory Card Not Supported',na=False)]

In [ ]:
smartphones_copy.loc[temp_df.index,'os']=np.nan

In [ ]:
temp_df=smartphones_copy[smartphones_copy['os'].str.contains('Memory Card',na=False)]

In [ ]:
temp_df.head(1)

In [ ]:
x=temp_df.iloc[:,10:12].shift(-1,axis=1).values

In [ ]:
x_index=temp_df.index

In [ ]:
smartphones_copy.loc[x_index,temp_df.columns[10:12]]=x

In [ ]:
smartphones_copy[smartphones_copy['os']=='Bluetooth']=np.nan

In [ ]:
smartphones_copy['os'].value_counts()

# **Tidiness**

In [ ]:
smartphones_copy.shape

In [ ]:
smartphones_copy.isnull().sum()

In [ ]:
smartphones_copy.head(1)

# model

In [ ]:
brand_name=smartphones_copy['model'].str.split().str.get(0)

In [ ]:
smartphones_copy.insert(1,'brand_name',brand_name)

In [ ]:
smartphones_copy.head(10)

In [ ]:
smartphones_copy['brand_name']=smartphones_copy['brand_name'].str.lower()

In [ ]:
smartphones_copy.head(2)

# sim

NFC (Near Field Communication) in mobile devices is a short-range wireless technology enabling secure, contactless communication for tasks like mobile payments, data exchange, and device pairing, requiring devices to be in close proximity.

# Has 5g

In [ ]:
smartphones_copy[smartphones_copy['sim'].str.contains('5G',na=False)]

In [ ]:
Has_5g=smartphones_copy['sim'].str.contains('5G',na=False)

# NFC

In [ ]:
smartphones_copy[smartphones_copy['sim'].str.contains('NFC',na=False)]

In [ ]:
has_Nfc=smartphones_copy['sim'].str.contains('NFC',na=False)

# ir blaster

In [ ]:
smartphones_copy['sim'].value_counts()

In [ ]:
smartphones_copy[smartphones_copy['sim'].str.contains(' IR Blaster',na=False)]

In [ ]:
Has_Ir=smartphones_copy['sim'].str.contains(' IR Blaster',na=False)

In [ ]:
smartphones_copy.insert(6,'has_5g',Has_5g)

In [ ]:
smartphones_copy.insert(7,'has_NFC',has_Nfc)

In [ ]:
smartphones_copy.insert(8,'has_IR_Blaster',Has_Ir)


In [ ]:
smartphones_copy.head(1)

# processor

In [ ]:
smartphones_copy['processor'].value_counts()

In [ ]:
Processor_name=  smartphones_copy['processor'].str.split(',').str.get(0)

In [ ]:
Num_cores=  smartphones_copy['processor'].str.split(',').str.get(1)

In [ ]:
Speed=  smartphones_copy['processor'].str.split(',').str.get(2)

In [ ]:
smartphones_copy.insert(10,'processor_name',Processor_name)

In [ ]:
smartphones_copy.insert(11,'num_cores',Num_cores)

In [ ]:
smartphones_copy.insert(12,'Speed',Speed)

In [ ]:
smartphones_copy.head(5)

In [ ]:
pd.set_option('display.max_rows',None)

In [ ]:
smartphones_copy['processor_name'].value_counts()

800 mAh Battery
2000 mAh Battery
1 GB RAM
4 MB RAM
Dual Core
1.77 MHz Processor
No 3G
Octa Core
No Wifi



In [ ]:
temp_df=smartphones_copy[smartphones_copy['processor_name'].str.strip().str.contains('Core',na=False)][['processor_name','num_cores','Speed']].shift(1,axis=1)

In [ ]:
smartphones_copy.loc[temp_df.index,['processor_name','num_cores','Speed']]=temp_df.values

In [ ]:
smartphones_copy['processor_name'].value_counts()

as per price these are feature phone

 1.77 MHz Processor

In [ ]:
smartphones_copy['processor_name'].value_counts()

In [ ]:
smartphones_copy[smartphones_copy['processor_name'].str.contains('28 nm',na=False)]

In [ ]:
smartphones_copy.loc[856,'processor_name']='Mediatek MT6739'

In [ ]:
smartphones_copy.loc[856]

In [ ]:
smartphones_copy['processor_name'].value_counts()

In [ ]:
Processor_name=smartphones_copy['processor_name'].str.split(' ').str.get(0).str.lower()

In [ ]:
smartphones_copy.insert(11,'processor_Brand',Processor_name)


In [ ]:
smartphones_copy.head(1)

In [ ]:
smartphones_copy['num_cores']=smartphones_copy['num_cores'].str.strip()

In [ ]:
smartphones_copy['num_cores'].value_counts()

In [ ]:
smartphones_copy['num_cores']=smartphones_copy['num_cores'].str.replace('Octa Core Processor','Octa Core').str.replace('Hexa Core Processor','Hexa Core')

In [ ]:
smartphones_copy['Speed']=smartphones_copy['Speed'].str.split().str.get(0).astype(float)

In [ ]:
smartphones_copy

In [ ]:
smartphones_copy.head(2)

In [ ]:
smartphones_copy['ram'].value_counts()

In [ ]:
import requests as re

In [ ]:
capacity=smartphones_copy['ram'].str.split(',').str[0].str.extract(r'(\d+\s?(TB|GB|MB|KB|B))')[0]

In [ ]:
smartphones_copy.insert(15,'RAM_capacity',capacity)

In [ ]:
capacity=smartphones_copy['ram'].str.split(',').str[1].str.extract(r'(\d+\s?(TB|GB|MB|KB|B))')[0]

In [ ]:
smartphones_copy.insert(16,'ROM_capacity',capacity)

In [ ]:
smartphones_copy.head(1)

In [ ]:
smartphones_copy['ram'].value_counts()

In [ ]:
smartphones_copy['RAM_capacity'].value_counts()

In [ ]:
smartphones_copy['ROM_capacity'].value_counts()

In [ ]:
smartphones_copy[smartphones_copy['ROM_capacity'].isin(['16 MB', '128 MB'])]

In [ ]:
smartphones_copy.drop(627,inplace=True)

In [ ]:
smartphones_copy.head(3)

# battery

In [ ]:
battery_capacity=smartphones_copy['battery'].str.split('with').str[0].str.extract(r'(\d+\s?(mAh))')[0]

In [ ]:
smartphones_copy.insert(18,'battery_capacity',battery_capacity)

In [ ]:
smartphones_copy['battery'][smartphones_copy['battery'].str.contains(' Fast Charging',na=False)]

In [ ]:
smartphones_copy['battery'].str.split('with').str[1]

In [ ]:
fast_charging=smartphones_copy['battery'].str.split('with').str[1]

In [ ]:
smartphones_copy.insert(19,'fast_charging',fast_charging)

In [ ]:
smartphones_copy.head(1)

# display

In [ ]:
screensize=smartphones_copy['display'].str.strip().str.split(',').str[0]

In [ ]:
smartphones_copy.insert(21,'screensize',screensize)

In [ ]:
resolution=smartphones_copy['display'].str.strip().str.split(',').str[1].str.strip().str.split(' ').str[0]

In [ ]:
smartphones_copy.insert(22,'resolution',resolution)

In [ ]:
smartphones_copy.info()

In [ ]:
refresh_rate=smartphones_copy['display'].str.strip().str.split(',').str.get(2).str.extract(r'(\d+\s?(Hz))')[0]

In [ ]:
smartphones_copy.insert(23,'refresh_rate',refresh_rate)

In [ ]:
smartphones_copy['refresh_rate'].fillna('60 Hz',inplace=True)

In [ ]:
smartphones_copy.head(2)

In [ ]:
smartphones_copy['refresh_rate'].value_counts()

#camera

In [ ]:
df=smartphones_copy[smartphones_copy['camera']=='Foldable Display, Dual Display']

In [ ]:
smartphones_copy.loc[df.index,'camera']='50 MP'

In [ ]:
def camera_extractor(text):
  if isinstance(text, float):  # अगर text NaN (float) है तो उसे empty string बना दो
       return ''
  if 'Quad' in text:
    return '4'
  elif 'Triple' in text:
    return '3'
  elif 'Dual' in text:
    return '2'
  elif 'Missing' in text:
    return 'Missing'
  else:
    return '1'

In [ ]:
num_of_rear_cameras=smartphones_copy['camera'].str.strip().str.split('&').str.get(0).apply(camera_extractor)

In [ ]:
smartphones_copy.insert(25,'num_of_rear_cameras',num_of_rear_cameras)

In [ ]:
num_of_front_cameras=smartphones_copy['camera'].str.strip().str.split('&').str.get(1).str.strip().fillna('Missing').apply(camera_extractor)

In [ ]:
smartphones_copy.insert(26,'num_of_front_cameras',num_of_front_cameras)

In [ ]:
rear_camera=smartphones_copy['camera'].str.strip().str.split('&').str.get(0)

In [ ]:
smartphones_copy.insert(27,'rear_camera',rear_camera)

In [ ]:
front=smartphones_copy['camera'].str.strip().str.split('&').str.get(1)

In [ ]:
smartphones_copy.insert(28,'front_camera',front)

In [ ]:
smartphones_copy.head(1)

In [ ]:
smartphones_copy[smartphones_copy['rear_camera'].isnull()]

In [ ]:
smartphones_copy.drop(486,inplace=True)

In [ ]:
smartphones_copy['rear_camera']=smartphones_copy['rear_camera'].str.strip().str.split('+').str.get(0).str.extract(r'(\d+)')

In [ ]:
smartphones_copy['rear_camera'].isnull().sum()

In [ ]:
smartphones_copy.shape

In [ ]:
smartphones_copy.head(2)

In [ ]:
smartphones_copy['front_camera']=smartphones_copy['front_camera'].str.strip().str.split('+').str.get(0).str.extract(r'(\d+)')

# card-> extended memory

In [ ]:
smartphones_copy['card'].fillna('Not Specified',inplace=True)

In [ ]:
smartphones_copy['extended_memory']=smartphones_copy['card'].apply(lambda x:'0' if 'Not' in x else x.split('upto')).str.get(-1).str.strip().str.replace('Memory Card Supported','Not Specified')

In [ ]:
temp_df=smartphones_copy[smartphones_copy['extended_memory'].str.contains('Android')]

In [ ]:
smartphones_copy.loc[temp_df.index,'extended_memory']='Not Specified'

In [ ]:
smartphones_copy['extended_memory'].value_counts()

In [ ]:
pd.set_option('display.max_columns', None)

In [ ]:
temp_df=smartphones_copy[smartphones_copy['extended_memory']=='Memory Card (Hybrid)']

In [ ]:
smartphones_copy.loc[temp_df.index,'extended_memory']='256 GB(Hybrid)'

In [ ]:
smartphones_copy

# os

In [ ]:
smartphones_copy['os'].value_counts()

In [ ]:
smartphones_copy['os'].fillna('android',inplace=True)

In [ ]:
def os_extractor(text):
    if isinstance(text, float):  # NaN values handle करने के लिए
        return "Not Specified"
    elif 'Android' in text:
        return 'android'
    elif 'iOS' in text:
        return 'ios'
    else:
        return "Other"

In [ ]:
smartphones_copy['os']=smartphones_copy['os'].apply(os_extractor)

In [ ]:
temp_df=smartphones_copy[smartphones_copy['os']=='Other']

In [ ]:
temp_df

In [ ]:
df=temp_df[temp_df['model'].str.contains('Gionee|Huawei', regex=True, case=False)]

In [ ]:
smartphones_copy.loc[df.index,'os']=' harmonyos'

In [ ]:
smartphones_copy.loc[temp_df.index,'os']='android'

# ADDITIONAL AMENDMENTS

In [ ]:
smartphones_copy['fast_charging']=smartphones_copy['fast_charging'].str.strip().str.split(' ').str.get(0).str.replace('Fast','0W').fillna('-1W')

In [ ]:
smartphones_copy['fast_charging']=smartphones_copy['fast_charging'].str.extract(r'(-?\d+)').astype(int)

In [ ]:
smartphones_copy['battery_capacity']=smartphones_copy['battery_capacity'].str.extract(r'(\d+)').fillna(0).astype(float)

In [ ]:
smartphones_copy['battery_capacity'].value_counts()

In [ ]:
smartphones_copy['RAM_capacity']=smartphones_copy['RAM_capacity'].str.extract(r'(\d+)').astype(float)

In [ ]:
smartphones_copy['ROM_capacity']=smartphones_copy['ROM_capacity'].str.extract(r'(\d+)').astype(float)

In [ ]:
smartphones_copy.head(2)

In [ ]:
smartphones_copy.shape

In [ ]:
smartphones_copy['refresh_rate']=smartphones_copy['refresh_rate'].str.extract(r'(\d+)').astype(int)

In [ ]:
smartphones_copy.info()

In [ ]:
smartphones_copy['price']=smartphones_copy['price'].astype(int)

In [ ]:
smartphones_copy.info()

In [ ]:
smartphones_copy.rename(columns={
    'index': 'index',
    'brand_name': 'brand_name',
    'model': 'model',
    'price': 'price (inr)',
    'rating': 'rating',
    'sim': 'sim',
    'has_5g': 'has_5g',
    'has_NFC': 'has_nfc',
    'has_IR_Blaster': 'has_ir_blaster',
    'processor': 'processor',
    'processor_name': 'processor_name',
    'processor_Brand': 'processor_brand',
    'num_cores': 'num_cores',
    'Speed': 'speed (ghz)',
    'ram': 'ram',
    'RAM_capacity': 'ram_capacity (gb)',
    'ROM_capacity': 'rom_capacity (gb)',
    'battery': 'battery',
    'battery_capacity': 'battery_capacity (mah)',
    'fast_charging': 'fast_charging (w)',
    'display': 'display',
    'screensize': 'screen_size (inches)',
    'resolution': 'resolution (pixels)',
    'refresh_rate': 'refresh_rate (hz)',
    'camera': 'camera',
    'num_of_rear_cameras': 'num_of_rear_cameras',
    'num_of_front_cameras': 'num_of_front_cameras',
    'rear_camera': 'rear_camera (mp)',
    'front_camera': 'front_camera (mp)',
    'card': 'card',
    'os': 'os',
    'extended_memory': 'extended_memory'
}, inplace=True)



In [ ]:
smartphones_copy.head(2)

In [ ]:
smartphones_copy['screen_size (inches)']=smartphones_copy['screen_size (inches)'].str.extract(r'(\d+\.\d+)').astype(float)

In [ ]:
smartphones_copy['resolution (pixels)']=smartphones_copy['resolution (pixels)'].str.strip().str.split('px').str.get(0)

In [ ]:
smartphones_copy.info()

In [ ]:
smartphones_copy['num_of_rear_cameras']=smartphones_copy['num_of_rear_cameras'].astype(int)

In [ ]:
smartphones_copy['num_of_front_cameras']=smartphones_copy['num_of_front_cameras'].str.replace('Missing','0').astype(int)

In [ ]:
smartphones_copy['rear_camera (mp)']=smartphones_copy['rear_camera (mp)'].astype(float)

In [ ]:
smartphones_copy['front_camera (mp)']=smartphones_copy['front_camera (mp)'].astype(float)


#Missing Values


In [ ]:
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt


In [ ]:
sns.boxenplot(x=smartphones_copy['rating'])

In [ ]:
sns.boxplot(x=smartphones_copy['fast_charging (w)'])

In [ ]:
plt.hist(smartphones_copy['battery_capacity (mah)'])

In [ ]:

bat=smartphones_copy[smartphones_copy['battery_capacity (mah)']==0]

In [ ]:
bat

In [ ]:
x=smartphones_copy[(smartphones_copy['brand_name'].str.contains('apple')) & (smartphones_copy['battery_capacity (mah)'] > 0)]['battery_capacity (mah)']

In [ ]:
smartphones_copy[smartphones_copy['brand_name'].str.contains('apple')]['battery_capacity (mah)'].describe()

In [ ]:
x.mean()

In [ ]:
median_battery = x.median()

In [ ]:
median_battery

In [ ]:
smartphones_copy.loc[bat.index,'battery_capacity (mah)']=median_battery

In [ ]:
smartphones_copy['battery_capacity (mah)'].value_counts()

In [ ]:
smartphones_copy.info()

In [ ]:
smartphones_copy[smartphones_copy['screen_size (inches)'].isnull()]

In [ ]:
x=smartphones_copy[~smartphones_copy['screen_size (inches)'].isnull()]

In [ ]:
x['screen_size (inches)'].mean()

In [ ]:
median_screen_size=x['screen_size (inches)'].median()

In [ ]:
smartphones_copy['screen_size (inches)'].fillna(median_screen_size,inplace=True)

# downloading Excel File

In [ ]:
smartphones_copy.to_excel('smartphone_cleaned_new.xlsx',index=False)
from google.colab import files
files.download('smartphone_cleaned_new.xlsx')

In [ ]:
smartphones_copy['rating'].fillna(smartphones_copy['rating'].mean(), inplace=True)


In [ ]:
smartphones_copy['rating'].isnull().sum()

In [ ]:
smartphones_copy.head(1)

In [ ]:
smartphones_copy.drop(columns=['index','sim','processor','ram','battery','display','camera','card'],inplace=True)

In [ ]:
smartphones_copy.info()

In [ ]:
smartphones_copy.to_csv('smartphone_cleaned_data.csv',index=False)
from google.colab import files
files.download('smartphone_cleaned_data.csv')